tests/integration/test_install.py

```diff
@@ -0,0 +1,140 @@
+# THIS FILE IS PART OF THE CYLC WORKFLOW ENGINE.
+# Copyright (C) NIWA & British Crown (Met Office) & Contributors.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Test cylc install."""
+
+import pytest
+from pathlib import Path
+
+from .test_scan import init_flows
+
+from cylc.flow.async_util import pipe
+from cylc.flow.scripts import scan
+from cylc.flow.workflow_files import WorkflowFiles
+from cylc.flow.scripts.install import (
+    InstallOptions,
+    install_cli
+)
+
+
+SRV_DIR = Path(WorkflowFiles.Service.DIRNAME)
+CONTACT = Path(WorkflowFiles.Service.CONTACT)
+RUN_N = Path(WorkflowFiles.RUN_N)
+INSTALL = Path(WorkflowFiles.Install.DIRNAME)
+
+INSTALLED_MSG = "INSTALLED {wfrun} from"
+WF_ACTIVE_MSG = '1 run of "{wf}" is already active:'
+BAD_CONTACT_MSG = "Bad contact file:"
+
+
+class graphql_query_mock:
+    """Context manager to mock network.scan.graphql_query().
+
+    (In the scripts.scan namepace, where it is used below).
+
+    """
```

Comment:
I couldn't figure out how to patch a decorated (`@pipe`) async function by standard means, so I resorted to this. Mocking and patching is still a bit of a black art, to me.

Reply:
I think we could do it be replacing the function in the place where it is being assembled into a pipe:

```python
def patch_graphql(monkeypatch):
    # define a mocked graphql pipe function
    @pipe
    def _graphql(*,args, **kwargs):
        pass

    # swap the graphql function that cylc.flow.scripts.scan sees for the patched version
    monkeypatch.setattr(
        'cylc.flow.scripts.graphql',
        _graphql,
    )

Reply:
Brilliant, with a couple of tweaks that worked, and it's cleaner than my approach so I've switched over. Cheers.